File: 2025/go/solution/day01.go
package solution

import (
	"log"
	"strconv"
	"strings"
)

type Day01 struct{}

func (d Day01) Part1(input string) string {
	// L -> ((100 - x) + cur) mod 100
	// R -> (x + cur) mod 100

	var (
		count = 0
		pos   = 50
	)

	lines := strings.SplitSeq(input, "\n")
	for line := range lines {
		// (L|R)(num)
		dir := line[0]
		num, _ := strconv.Atoi(line[1:])
		switch dir {
		case 'L':
			pos = 100 - num + pos
		case 'R':
			pos = num + pos
		default:
			log.Fatalf("unknown first character: %c\n", dir)
		}
		pos %= 100
		if pos == 0 {
			count += 1
		}
	}

	return strconv.Itoa(count)
}

func (d Day01) Part2(input string) string {
	// Count every time 0 is passed
	var (
		count = 0
		pos   = 50
	)

	lines := strings.SplitSeq(input, "\n")
	for line := range lines {
		dir := line[0]
		num, _ := strconv.Atoi(line[1:])
		switch dir {
		case 'L':
			// Don't go negative
			if pos == 0 {
				pos = 100
			}
			for _ = range num {
				pos -= 1
				if pos == 0 {
					count += 1
					pos = 100
				}
			}
		case 'R':
			// Don't go above 100
			if pos == 100 {
				pos = 0
			}
			for _ = range num {
				pos += 1
				if pos == 100 {
					count += 1
					pos = 0
				}
			}
		default:
			log.Fatalf("unknown first character: %c\n", dir)
		}
	}
	return strconv.Itoa(count)
}

File: 2025/rust/src/solution/day01.rs
use crate::solution::Solution;

pub struct Day01;

fn parse(input: &str) -> Vec<(&str, i32)> {
    input
        .lines()
        .map(|line| {
            let (dir, turns) = line.split_at(1);
            let turns = turns.parse::<i32>().expect("should be valid integer");
            (dir, turns)
        })
        .collect::<_>()
}

fn modulo(a: i32, b: i32) -> i32 {
    (a % b + b) % b
}

impl Solution for Day01 {
    fn part1(&self, input: &str) -> String {
        let mut pos: i32 = 50;
        let mut count = 0;

        let ops = parse(input);
        for (dir, turns) in ops {
            match dir {
                "L" => pos -= turns,
                "R" => pos += turns,
                c => unreachable!("got prefix: {c}"),
            }
            pos = modulo(pos, 100);
            if pos == 0 {
                count += 1;
            }
        }

        format!("{count}")
    }

    fn part2(&self, input: &str) -> String {
        todo!()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const INPUT: &str = r"L68
L30
R48
L5
R60
L55
L1
L99
R14
L82";

    #[test]
    fn test_part1() {
        let expected = "3".to_string();
        let actual = Day01.part1(INPUT);
        assert_eq!(expected, actual);
    }
}
